clippy_lints/src/types.rs

```diff
@@ -1210,21 +1210,26 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Casts {
             let (cast_from, cast_to) = (cx.tables.expr_ty(ex), cx.tables.expr_ty(expr));
             lint_fn_to_numeric_cast(cx, expr, ex, cast_from, cast_to);
             if let ExprKind::Lit(ref lit) = ex.kind {
-                if let LitKind::Int(n, _) = lit.node {
-                    if cast_to.is_floating_point() {
+                if_chain! {
+                    if let LitKind::Int(n, _) = lit.node;
+                    if let Some(src) = snippet_opt(cx, lit.span);
+                    if cast_to.is_floating_point();
+                    then {
                         let from_nbits = 128 - n.leading_zeros();
                         let to_nbits = fp_ty_mantissa_nbits(cast_to);
-                        if from_nbits != 0 && to_nbits != 0 && from_nbits <= to_nbits {
-                            span_lint_and_sugg(
-                                cx,
-                                UNNECESSARY_CAST,
-                                expr.span,
-                                &format!("casting integer literal to `{}` is unnecessary", cast_to),
-                                "try",
-                                format!("{}_{}", n, cast_to),
-                                Applicability::MachineApplicable,
-                            );
-                            return;
+                        if let Some(num_lit) = NumericLiteral::from_lit_kind(&src, &lit.node) {
+                            if from_nbits != 0 && to_nbits != 0 && from_nbits <= to_nbits && num_lit.is_decimal() {
```

Comment:
You can put this also in the `if_chain!`. The `if_chain` macro also allows `let`s in it.

Also you can move the `let` binding down one line.

Reply:
Done.